File: app/src/main/java/net/luanwilliam/cookbook/CadastroReceita.java
package net.luanwilliam.cookbook;

import androidx.appcompat.app.AppCompatActivity;
import androidx.loader.content.AsyncTaskLoader;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.AsyncNotedAppOp;
import android.app.DatePickerDialog;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.Toast;

import net.luanwilliam.cookbook.DAO.DAO;
import net.luanwilliam.cookbook.DAO.ReceitaDatabase;
import net.luanwilliam.cookbook.controle.Controle;
import net.luanwilliam.cookbook.model.Categoria;
import net.luanwilliam.cookbook.model.Receita;

import java.sql.SQLOutput;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class CadastroReceita extends AppCompatActivity {
    private Spinner spnGrupos;
    private EditText nome, descricao, ingredientes, modo_preparo;
    private RadioGroup grupoClassificacao;
    private CheckBox check;
    public String avaliacao;
    private Button limpar, cadastrar;
    private ImageButton btnCadCat, btnDelCat, btnAltCat;
    private List<Categoria> listaCategoria;

    private ArrayAdapter<Categoria> adapterCategorias;

 //   DAO dao = new DAO(this);
    private long id = 0;
    public static final int ALTERAR = 1;
    public static final int NOVO = 0;
    public static final String MODO = "MODO";
    public static final String ID_ALTERAR = "ID";
    private Receita rec ;
    private int modo;

    private DatePickerDialog datePickerDialog;
    private Button btnDatePicker;

    private String datePicked = getHoje();

    private long idAlterar;






    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cadastro_receita);
        ativarElementos();




        btnDatePicker = findViewById(R.id.dataPicker);
        btnDatePicker.setText(getHoje());
        pushSpinner();
        iniciarDatePicker();


        Intent intent = getIntent();
        Bundle bundle = intent.getExtras();

        modo = bundle.getInt(MODO, NOVO);

     //   rButton(grupoClassificacao);

        if(modo == ALTERAR){
            setTitle("Alterar ");
            ativarElementos();
            subirDados();

             idAlterar = bundle.getLong("id");


            cadastrar.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    atualizar();
                }
            });



        }else {
            setTitle("Nova Receita");

            cadastrar.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    cadastro(NOVO, -1);
                }
            });




           /* AsyncTask.execute(new Runnable() {
                @Override
                public void run() {
                    ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);

                }
            }); */

        }







                limpar.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        limparTela();
                    }
                });

                btnCadCat.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Bundle bundle = new Bundle();
                        bundle.putInt(MODO, NOVO);

                        Intent intencao = new Intent(CadastroReceita.this, CadastroCategoria.class);
                        intencao.putExtras(bundle);
                        startActivity(intencao);
                    }
                });



               /* btnAltCat.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Spinner x = findViewById(R.id.grupo_spinner);
                         String  nomec = (String) x.getSelectedItem();
                        System.out.println( "teste 2 nomec "+nomec);
                        String nomeCat = (String) spnGrupos.getSelectedItem();
                        ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);
                        long id = db.categoriaDAO().buscaByNome(nomeCat);
                        Intent intent = new Intent();
                        Bundle bundle = new Bundle();
                        bundle.putLong(ID_ALTERAR, id);
                        bundle.putInt(MODO,ALTERAR);


                    }
                }); */





    }

    private void pushSpinner(){


        AsyncTask.execute( () -> {
            ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);
            listaCategoria = db.categoriaDAO().queryAll();
            // db.close();

            //  lista.add("Selecione a categoria");
           // for (Categoria c : listaCategoria) {
            //    lista.add(c.getNome());
           // }

            CadastroReceita.this.runOnUiThread(() -> {
                adapterCategorias = new ArrayAdapter<>(this,  android.R.layout.simple_list_item_1, listaCategoria);

              //  adaptador = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, lista);
                spnGrupos.setAdapter(adapterCategorias);
            });

        });
    }

    private void popularSpinner(){
        ReceitaDatabase bd = ReceitaDatabase.getDatabase(this);
        listaCategoria = bd.categoriaDAO().queryAll();
        bd.close();

        List<String> lista = new ArrayList<>();
       // lista.add("Selecione a categoria");
        for(Categoria c : listaCategoria){
            lista.add(c.getNome());
        }


        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_2 , lista);
        spnGrupos.setAdapter(adapter);
    }



    public void ativarElementos(){
        this.nome = findViewById(R.id.text_nome);
        this.descricao = findViewById(R.id.text_descricao);
        this.ingredientes = findViewById(R.id.ingredientes);
        this.modo_preparo = findViewById(R.id.text_preparo);
        this.grupoClassificacao = findViewById(R.id.radioGroup);
        this.check = findViewById(R.id.checkboxFez);
        rButton(grupoClassificacao);
        this.spnGrupos = findViewById(R.id.grupo_spinner);
        pushSpinner();
        this.limpar = findViewById(R.id.btnLimpar);
        this.cadastrar = findViewById(R.id.btnCadastrar);
        this.btnCadCat = findViewById(R.id.btnCadastrarCategoria);
        this.btnDelCat = findViewById(R.id.btnDeleteCategoria);
        this.btnAltCat = findViewById(R.id.btnAlterarCategoria);
        this.btnDatePicker = findViewById(R.id.dataPicker);
        this.btnDatePicker.setText(getHoje());


    }



    public void rButton(RadioGroup grupo){


        grupo.setOnCheckedChangeListener( new RadioGroup.OnCheckedChangeListener(){

            /**
             * <p>Called when the checked radio button has changed. When the
             * selection is cleared, checkedId is -1.</p>
             *
             * @param group     the group in which the checked radio button has changed
             * @param checkedId the unique identifier of the newly checked radio button
             */
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                RadioButton button = (RadioButton) group.findViewById(checkedId);
                String res = button.getText().toString();
                avaliacao = res;
            }
        });
    }

    private void limparTela(){
        this.nome.setText("");
        this.descricao.setText("");
        this.ingredientes.setText("");
        this.modo_preparo.setText("");
      //  grupoClassificacao.clearCheck();
       // popularSpinner(dao.carregarCatBD());




    }

    public static void alterarReceita(Activity act , int requestCode, Receita rec){
        Intent intent = new Intent(act, CadastroReceita.class);

        intent.putExtra(MODO, ALTERAR);
        intent.putExtra(ID_ALTERAR, rec.getIdrec());

        act.startActivityForResult(intent, requestCode);

    }

    public void cadastro(int valor, long id){
        ativarElementos();
        System.out.println(" antes testes ");
        String nomeC = nome.getText().toString();
        String desc = descricao.getText().toString();
        String ing = ingredientes.getText().toString();
        String modo = modo_preparo.getText().toString();
        String cate = spnGrupos.getSelectedItem().toString();
        String data = getDate();
        System.out.println("teste 1 data "+data);
        rButton(grupoClassificacao);

        boolean checked = false;
        if(check.isChecked()){
            checked = true;
        }
       // System.out.println("teste cate "+cate);

        System.out.println("testes ");

        System.out.println("nome cadastros " + nomeC);
        System.out.println("desc "+desc);
        System.out.println("Categoria "+cate);
        System.out.println("Avaliaçao "+ avaliacao);
        System.out.println("teste check "+checked);
        Categoria cat = (Categoria)spnGrupos.getSelectedItem();
        System.out.println("teste categoria "+ cat.getNome());



        ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);

        if(valor == NOVO){
           // Categoria c = encontrarCat(cate);
          //  System.out.println("testar categoria "+c.getNome());
            Receita rec = new Receita(nomeC, desc, ing, modo, checked,avaliacao, cat.getIdcat() );
            rec.setData(data);
            AsyncTask.execute( () ->{
                //ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);
                db.receitaDAO().insert(rec);

              //  db.close();


            });

            Toast.makeText(CadastroReceita.this, "Cadastrado com sucesso", Toast.LENGTH_LONG).show();



        }

        Intent intent = new Intent(CadastroReceita.this, MainActivity.class);
        startActivity(intent);
        finishActivity(RESULT_OK);



    }


    public void atualizar(){


        System.out.println(" antes testes ");
        String nomeC = nome.getText().toString();
        String desc = descricao.getText().toString();
        String ing = ingredientes.getText().toString();
        String modo = modo_preparo.getText().toString();
        String cate = spnGrupos.getSelectedItem().toString();
        String data = getDate();
        System.out.println("teste 1 data "+data);
        rButton(grupoClassificacao);

        boolean checked = false;
        if(check.isChecked()){
            checked = true;
        }
        // System.out.println("teste cate "+cate);

        System.out.println("testes ");

        System.out.println("nome cadastros " + nomeC);
        System.out.println("desc "+desc);
        System.out.println("Categoria "+cate);
        System.out.println("Avaliaçao "+ avaliacao);
        System.out.println("teste check "+checked);
        Categoria cat = (Categoria)spnGrupos.getSelectedItem();
        System.out.println("teste categoria "+ cat.getNome());



        ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);

        Receita upRec = new Receita(nomeC, desc, ing, modo, checked,avaliacao, cat.getIdcat());
        upRec.setIdrec(idAlterar);
        upRec.setData(data);


        AsyncTask.execute( () ->{

            db.receitaDAO().update(upRec);
        });

        Toast.makeText(CadastroReceita.this, "Alterado com Sucesso", Toast.LENGTH_LONG).show();

        Intent intent = new Intent(CadastroReceita.this, MainActivity.class);
        startActivity(intent);
        finishActivity(RESULT_OK);



    }

    public void subirDados(){


        Intent intent = getIntent();

        Bundle param = intent.getExtras();
        long id = param.getLong("id");

        String nome = param.getString("nome");
        String desc = param.getString("descricao");
        String ingr = param.getString("ingredientes");
        String modoP = param.getString("modo_preparo");
        String data = param.getString("data");


        this.nome.setText(nome);
        this.descricao.setText(desc);
        this.ingredientes.setText(ingr);
        this.modo_preparo.setText(modoP);

    }

    public Categoria encontrarCat(String nome){


    AsyncTask.execute( () -> {
        ReceitaDatabase db = ReceitaDatabase.getDatabase(CadastroReceita.this);
        listaCategoria = db.categoriaDAO().queryAll();

    });

        for(Categoria c : listaCategoria){
            if(c.getNome().equals(nome)){
                return c;
            }
        }
        return null;



    }

    public void mostrarDatePicker(View view) {
        datePickerDialog.show();


    }

    public void iniciarDatePicker(){
        DatePickerDialog.OnDateSetListener dateSetListener = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                month += 1;


                String data = makeDateString(dayOfMonth, month, year);

                btnDatePicker.setText(data);
                setDate(data);

            }
        };
        Calendar calendar = Calendar.getInstance();
        int ano = calendar.get(Calendar.YEAR);
        int dia = calendar.get(Calendar.DAY_OF_MONTH);
        int mes = calendar.get(Calendar.MONTH);

        int estilo = AlertDialog.THEME_HOLO_LIGHT;

        datePickerDialog =  new DatePickerDialog(this, estilo, dateSetListener, dia, mes, ano);

    }

    private String makeDateString(int dayOfMonth, int month, int year) {

        return dayOfMonth + "/" + getMonthFormat(month )+ "/"+ year;
    }

    private String getMonthFormat(int month) {

        if(month == 1){
            return "JAN";
        }
        if(month == 2){
            return "FEV";
        }if(month == 3){
            return "MAR";
        }if(month == 4){
            return "ABR";
        }if(month == 5){
            return "MAIO";
        }if(month == 6){
            return "JUN";
        }if(month == 7){
            return "JUL";
        }if(month == 8){
            return "AGOS";
        }if(month == 9){
            return "SET";
        }if(month == 10){
            return "OUT";
        }if(month == 11){
            return "NOV";
        }if(month == 12){
            return "DEZ";
        }
        return "JAN";


    }


    private String getHoje() {
        Calendar calendar = Calendar.getInstance();
        int ano = calendar.get(Calendar.YEAR);
        int dia = calendar.get(Calendar.DAY_OF_MONTH);
        int mes = calendar.get(Calendar.MONTH);
        mes += 1;
        return makeDateString(dia, mes, ano);


    }

    public void setDate(String date){
        datePicked = date;
    }
    public String getDate(){
        return datePicked;
    }

}

File: app/src/main/kotlin/net/luanwilliam/cookbook/CadastroReceita.kt
package net.luanwilliam.cookbook

import android.app.AlertDialog
import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.ImageButton
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.Spinner
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import net.luanwilliam.cookbook.dao.ReceitaDatabase
import net.luanwilliam.cookbook.model.Categoria
import net.luanwilliam.cookbook.model.Receita
import java.util.*

class CadastroReceita : AppCompatActivity() {
    private var spnGrupos: Spinner? = null
    private var nome: EditText? = null
    private var descricao: EditText? = null
    private var ingredientes: EditText? = null
    private var modo_preparo: EditText? = null
    private var grupoClassificacao: RadioGroup? = null
    private var check: CheckBox? = null
    var avaliacao: String? = ""
    private var limpar: Button? = null
    private var cadastrar: Button? = null
    private var btnCadCat: ImageButton? = null
    private var btnDelCat: ImageButton? = null
    private var btnAltCat: ImageButton? = null
    private var listaCategoria: List<Categoria>? = null
    private var adapterCategorias: ArrayAdapter<Categoria>? = null

    private var modo = 0
    private var datePickerDialog: DatePickerDialog? = null
    private var btnDatePicker: Button? = null
    var date = hoje
    private var idAlterar: Long = 0

    companion object {
        const val ALTERAR = 1
        const val NOVO = 0
        const val MODO = "MODO"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cadastro_receita)
        ativarElementos()
        iniciarDatePicker()
        btnDatePicker = findViewById(R.id.dataPicker)
        pushSpinner()
        val intent = intent
        val bundle = intent.extras
        modo = bundle!!.getInt(MODO, NOVO)

        if (modo == ALTERAR) {
            title = "Alterar "
            ativarElementos()
            subirDados()
            idAlterar = bundle.getLong("id")
            cadastrar!!.setOnClickListener { atualizar() }
        } else {
            title = "Nova Receita"
            cadastrar!!.setOnClickListener { cadastro() }
        }
        limpar!!.setOnClickListener { limparTela() }
        btnCadCat!!.setOnClickListener {
            val otherBundle = Bundle()
            otherBundle.putInt(MODO, NOVO)
            val intencao = Intent(this@CadastroReceita, CadastroCategoria::class.java)
            intencao.putExtras(otherBundle)
            startActivity(intencao)
        }
    }

    private fun pushSpinner() {
        AsyncTask.execute {
            val db = ReceitaDatabase.getDatabase(this@CadastroReceita)
            listaCategoria = db!!.categoriaDAO().queryAll() as List<Categoria>?
            runOnUiThread {
                adapterCategorias = ArrayAdapter(this, android.R.layout.simple_list_item_1, listaCategoria!!)
                spnGrupos!!.adapter = adapterCategorias
            }
        }
    }

    private fun popularSpinner() {
        val bd = ReceitaDatabase.getDatabase(this)
        listaCategoria = bd!!.categoriaDAO().queryAll() as List<Categoria>?
        bd.close()
        val lista: MutableList<String> = ArrayList()
        for (c in listaCategoria!!) {
            lista.add(c.nome)
        }
        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_2, lista)
        spnGrupos!!.adapter = adapter
    }

    private fun ativarElementos() {
        nome = findViewById(R.id.text_nome)
        descricao = findViewById(R.id.text_descricao)
        ingredientes = findViewById(R.id.ingredientes)
        modo_preparo = findViewById(R.id.text_preparo)
        grupoClassificacao = findViewById(R.id.radioGroup)
        check = findViewById(R.id.checkboxFez)
        rButton(grupoClassificacao)
        spnGrupos = findViewById(R.id.grupo_spinner)
        pushSpinner()
        limpar = findViewById(R.id.btnLimpar)
        cadastrar = findViewById(R.id.btnCadastrar)
        btnCadCat = findViewById(R.id.btnCadastrarCategoria)
        btnDelCat = findViewById(R.id.btnDeleteCategoria)
        btnAltCat = findViewById(R.id.btnAlterarCategoria)
        btnDatePicker = findViewById(R.id.dataPicker)
        btnDatePicker!!.text = hoje
    }

    private fun rButton(grupo: RadioGroup?) {
        grupo!!.setOnCheckedChangeListener { group, checkedId ->
            val button = group.findViewById<View>(checkedId) as RadioButton
            val res = button.text.toString()
            avaliacao = res
        }
    }

    private fun limparTela() {
        nome!!.setText("")
        descricao!!.setText("")
        ingredientes!!.setText("")
        modo_preparo!!.setText("")
    }

    private fun cadastro() {
        ativarElementos()
        println(" antes testes ")
        val nomeC = nome!!.text.toString()
        val desc = descricao!!.text.toString()
        val ing = ingredientes!!.text.toString()
        val modo = modo_preparo!!.text.toString()
        val cate = spnGrupos!!.selectedItem.toString()
        val data = date
        println("teste 1 data $data")
        rButton(grupoClassificacao)
        var checked = false
        if (check!!.isChecked) {
            checked = true
        }
        println("testes ")
        println("nome cadastros $nomeC")
        println("desc $desc")
        println("Categoria $cate")
        println("Avaliaçao $avaliacao")
        println("teste check $checked")
        val cat = spnGrupos!!.selectedItem as Categoria
        println("teste categoria " + cat.nome)
        val db = ReceitaDatabase.getDatabase(this@CadastroReceita)

        val rec = Receita(nomeC, desc, ing, modo, checked, avaliacao ?: "", cat.idcat)
        rec.data = data
        AsyncTask.execute {
            db!!.receitaDAO()!!.insert(rec)
        }
        Toast.makeText(this@CadastroReceita, "Cadastrado com sucesso", Toast.LENGTH_LONG).show()

        val intent = Intent(this@CadastroReceita, MainActivity::class.java)
        startActivity(intent)
        finishActivity(RESULT_OK)
    }

    fun atualizar() {
        println(" antes testes ")
        val nomeC = nome!!.text.toString()
        val desc = descricao!!.text.toString()
        val ing = ingredientes!!.text.toString()
        val modo = modo_preparo!!.text.toString()
        val cate = spnGrupos!!.selectedItem.toString()
        val data = date
        println("teste 1 data $data")
        rButton(grupoClassificacao)
        var checked = false
        if (check!!.isChecked) {
            checked = true
        }

        println("testes ")
        println("nome cadastros $nomeC")
        println("desc $desc")
        println("Categoria $cate")
        println("Avaliaçao $avaliacao")
        println("teste check $checked")
        val cat = spnGrupos!!.selectedItem as Categoria
        println("teste categoria " + cat.nome)
        val db = ReceitaDatabase.getDatabase(this@CadastroReceita)
        val upRec = Receita(nomeC, desc, ing, modo, checked, avaliacao!!, cat.idcat)
        upRec.idrec = idAlterar
        upRec.data = data
        AsyncTask.execute { db!!.receitaDAO()!!.update(upRec) }
        Toast.makeText(this@CadastroReceita, "Alterado com Sucesso", Toast.LENGTH_LONG).show()
        val intent = Intent(this@CadastroReceita, MainActivity::class.java)
        startActivity(intent)
        finishActivity(RESULT_OK)
    }

    private fun subirDados() {
        val intent = intent
        val param = intent.extras
        val id = param!!.getLong("id")
        val nome = param.getString("nome")
        val desc = param.getString("descricao")
        val ingr = param.getString("ingredientes")
        val modoP = param.getString("modo_preparo")
        val data = param.getString("data")
        this.nome!!.setText(nome)
        descricao!!.setText(desc)
        ingredientes!!.setText(ingr)
        modo_preparo!!.setText(modoP)
    }

    fun mostrarDatePicker(view: View?) {
        datePickerDialog!!.show()
    }

    private fun iniciarDatePicker() {
        btnDatePicker!!.text = date
        val dateSetListener = OnDateSetListener { view, year, month, dayOfMonth ->
            var month = month
            month += 1
            val data = makeDateString(dayOfMonth, month, year)
            btnDatePicker!!.text = data
            //this.datePicked = data //TODO: Is this needed ?
        }

        val calendar = Calendar.getInstance()
        val ano = calendar[Calendar.YEAR]
        val dia = calendar[Calendar.DAY_OF_MONTH]
        val mes = calendar[Calendar.MONTH]
        val estilo = AlertDialog.THEME_HOLO_LIGHT
        datePickerDialog = DatePickerDialog(this, estilo, dateSetListener, dia, mes, ano)
    }

    private fun makeDateString(dayOfMonth: Int, month: Int, year: Int): String {
        return dayOfMonth.toString() + "/" + getMonthFormat(month) + "/" + year
    }

    private fun getMonthFormat(month: Int): String { //TODO: :(
        if (month == 1) {
            return "JAN"
        }
        if (month == 2) {
            return "FEV"
        }
        if (month == 3) {
            return "MAR"
        }
        if (month == 4) {
            return "ABR"
        }
        if (month == 5) {
            return "MAIO"
        }
        if (month == 6) {
            return "JUN"
        }
        if (month == 7) {
            return "JUL"
        }
        if (month == 8) {
            return "AGOS"
        }
        if (month == 9) {
            return "SET"
        }
        if (month == 10) {
            return "OUT"
        }
        if (month == 11) {
            return "NOV"
        }
        return if (month == 12) {
            "DEZ"
        } else "JAN"
    }

    private val hoje: String
        get() {
            val calendar = Calendar.getInstance()
            val ano = calendar[Calendar.YEAR]
            val dia = calendar[Calendar.DAY_OF_MONTH]
            var mes = calendar[Calendar.MONTH]
            mes += 1
            return makeDateString(dia, mes, ano)
        }
}